lang/wasm.cc

```diff
@@ -19,190 +17,97 @@
+static void init_abstract_arg(const abstract_type& t, const bytes_opt& param, wasmtime::ValVec& argv, wasmtime::Store& store, wasmtime::Instance& instance) {
         // set up exported memory's underlying buffer,
         // `memory` is required to be exported in the WebAssembly module
-        auto memory_export = instance.get(store, "memory");
-        if (!memory_export) {
-            throw wasm::exception("memory export not found - please export `memory` in the wasm module");
-        }
-        auto memory = std::get<wasmtime::Memory>(*memory_export);
-        uint8_t* data = memory.data(store).data();
-        size_t mem_size = memory.size(store) * WASM_PAGE_SIZE;
-        int32_t serialized_size = param ? param->size() : 0;
-        if (serialized_size > std::numeric_limits<int32_t>::max()) {
+        auto memory = wasmtime::get_memory(instance, store);
+        size_t mem_size = memory->size(store) * WASM_PAGE_SIZE;
+        if (param && param->size() > std::numeric_limits<int32_t>::max()) {
             throw wasm::exception(format("Serialized parameter is too large: {} > {}", param->size(), std::numeric_limits<int32_t>::max()));
         }
-        switch (uint32_t abi_ver = get_abi(instance, store, data)) {
-            case 1: {
-                auto grown = memory.grow(store, 1 + (sizeof(int32_t) + serialized_size - 1) / WASM_PAGE_SIZE); // for fitting serialized size + the buffer itself
-                if (!grown) {
-                    throw wasm::exception(format("Failed to grow wasm memory to {}: {}", serialized_size, grown.err().message()));
+        int32_t serialized_size = param ? param->size() : 0;
+        if (param) {
+            switch (uint32_t abi_ver = wasmtime::get_abi(instance, store, *memory)) {
+                case 1: {
+                    auto pre_grow = memory->grow(store, 1 + (serialized_size - 1) / WASM_PAGE_SIZE);
+                    mem_size = pre_grow * WASM_PAGE_SIZE;
+                    break;
                 }
-                mem_size = std::move(grown).unwrap() * WASM_PAGE_SIZE;
-                break;
-            }
-            case 2: {
-                auto malloc_func = import_func(instance, store, "_scylla_malloc");
-                import_func(instance, store, "_scylla_free");
-                auto size = call_func(store, malloc_func, {int32_t(sizeof(int32_t) + serialized_size)});
-                mem_size = size.i32();
-                break;
+                case 2: {
+                    auto malloc_func = wasmtime::create_func(instance, store, "_scylla_malloc");
+                    wasmtime::create_func(instance, store, "_scylla_free");
+                    auto argv = wasmtime::get_val_vec();
+                    argv->push_i32(serialized_size);
+                    auto rets = wasmtime::get_val_vec();
+                    rets->push_i32(0);
+
+                    auto fut = wasmtime::get_func_future(store, *malloc_func, *argv, *rets);
+                    while(!fut->resume());
```

Comment:
Code style: space after while, it isn't a function.

Also a comment indicating that we need to yield here.

If we set need_thread, then it may be enough to call seastar::thread::maybe_yield() here.

Reply:
I can modify the `operator()` in the `init(_nullable)_arg_visitor` so that it returns a future, then we can use the `maybe_yield` like when executing the UDF

Reply:
If this is just pushing arguments, maybe unneeded. But comments are definitely needed.